src/Handler/UnionHandler.php

```diff
@@ -47,46 +48,91 @@ public function serializeUnion(
         mixed $data,
         array $type,
         SerializationContext $context
-    ) {
-        return $this->matchSimpleType($data, $type, $context);
+    ): mixed {
+        if ($this->isPrimitiveType(gettype($data))) {
+            return $this->matchSimpleType($data, $type, $context);
+        } else {
+            $resolvedType = [
+                'name' => get_class($data),
+                'params' => [],
+            ];
+
+            return $context->getNavigator()->accept($data, $resolvedType);
+        }
     }
 
-    public function deserializeUnion(DeserializationVisitorInterface $visitor, mixed $data, array $type, DeserializationContext $context)
+    public function deserializeUnion(DeserializationVisitorInterface $visitor, mixed $data, array $type, DeserializationContext $context): mixed
     {
         if ($data instanceof \SimpleXMLElement) {
             throw new RuntimeException('XML deserialisation into union types is not supported yet.');
         }
 
-        return $this->matchSimpleType($data, $type, $context);
-    }
-
-    private function matchSimpleType(mixed $data, array $type, Context $context)
-    {
-        $dataType = $this->determineType($data, $type, $context->getFormat());
-        $alternativeName = null;
-
-        if (isset(static::$aliases[$dataType])) {
-            $alternativeName = static::$aliases[$dataType];
+        $finalType = null;
+        if (2 === count($type['params'])) {
+            if (!is_array($type['params'][0]) || !array_key_exists('name', $type['params'][0])) {
+                $lookupField = $type['params'][0];
+                $unionMap = $type['params'][1];
+
+                if (!array_key_exists($lookupField, $data)) {
+                    throw new NonVisitableTypeException('Union Discriminator Field \'' . $lookupField . '\' not found in data');
+                }
+
+                $lkup = $data[$lookupField];
+                if (!empty($unionMap)) {
+                    if (array_key_exists($lkup, $unionMap)) {
+                        $finalType = [
+                            'name' => $unionMap[$lkup],
+                            'params' => [],
+                        ];
+                    } else {
+                        throw new NonVisitableTypeException('Union Discriminator Map does not contain key \'' . $lkup . '\'');
+                    }
+                } else {
```

Comment:
this `else` branch still exposes the ability to pass any class name in the json and later that class wiill be instantiated.  I think that this should be removed